packages/client-api-schema/src/types.ts

```diff
@@ -181,20 +181,37 @@ export type GetBudgetResponse = JsonRpcResponse<SiteBudget | {}>;
 
 export type ApproveBudgetAndFundRequest = JsonRpcRequest<'ApproveBudgetAndFund', BudgetRequest>;
 export type ApproveBudgetAndFundResponse = JsonRpcResponse<SiteBudget>;
+
 // Notifications
 export type ChannelProposedNotification = JsonRpcNotification<'ChannelProposed', ChannelResult>;
 export type ChannelUpdatedNotification = JsonRpcNotification<'ChannelUpdated', ChannelResult>;
 export type ChannelClosingNotification = JsonRpcNotification<'ChannelClosed', ChannelResult>;
 export type MessageQueuedNotification = JsonRpcNotification<'MessageQueued', Message>;
 export type BudgetUpdatedNotification = JsonRpcNotification<'BudgetUpdated', SiteBudget>;
 
+// these notifications come *from* the wallet, which is not how JSON-RPC works
 export type Notification =
   | ChannelProposedNotification
   | ChannelUpdatedNotification
   | ChannelClosingNotification
   | BudgetUpdatedNotification
   | MessageQueuedNotification;
 
+type FilterByMethod<T, Method> = T extends {method: Method} ? T : never;
+
+export type NotificationType = {
+  [T in Notification['method']]: [FilterByMethod<Notification, T>['params']];
+};
```

Comment:
Constructs a type that is a lookup from the allowed `method`s to the corresponding `params`.